Revert "Clear" Button from <button> to <a> Tag for Improved Theme Compatibility while maintaining accessibility improvements #54309

imanish003 opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

In WooCommerce 9.5, PR #50183 changed the "Clear" button on single product pages for variable products from an `<a>` tag to a `<button>` tag. While this improved accessibility, it is causing (widespread) visual issues, the Clear button now appears more prominent than the Add to Cart button. After internal discussion on Slack (p1736320650381359-slack-C02FL3X7KR6), it was decided to change the Clear button back to an `<a>` tag.

In summary:
- Replaced `<button>` tag with `<a>` tag and added `role="button"` for accessibility.
- Added keyboard event handling (Enter/Space) to match button behavior i.e. user will be able to use Enter or Space to clear the selection.
- Most of changes made in PR #50183 are preserved. For example, clearing attribute values will add a screen reader text "Your selection has been reset. Please select some product options before adding this product to your cart."

![image](https://github.com/user-attachments/assets/9759269d-d4f8-4bfe-b521-97781726bf65)

It was also raised on woo-dev slack channel (p1736165110941609-slack-C7U3Y3VMY).

Closes #54085

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

#### Verify that the "Clear" button is an `<a>` tag

1. Go to frontend of any variable product. For example, Hoodie (in case you are using sample products)
2. Select some variation attributes. For example, select "Red" for "Color".
3. Verify that the "Clear" button is an `<a>` tag. See before vs this PR screenshot below.
![image](https://github.com/user-attachments/assets/9759269d-d4f8-4bfe-b521-97781726bf65)

#### Scenario: Testing for regression

> [!NOTE]
> This PR reverts the change from PR #50183 that converted the "Clear" link from an `<a>` tag back to a `<button>` tag. We need to verify that this reversion maintains the same functionality and accessibility improvements that were introduced in #50183.

1. Activate screen reader of your choice.
	- Tip: You can use shortcut “Cmd + F5” to enable VoiceOver on Mac.
2. Go to frontend of any variable product. For example, Hoodie (in case you are using sample products)
3. Tab to variation attributes, use your arrows and space bar to make a selection. For example, select "Red" for "Color".
4. Tab to the "Clear" button.
5. Verify that you can use either Enter or Spacebar to clear the selection.
6. Verify that you hear a message "Your selection has been reset. Please select some product options before adding this product to your cart."

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Fix: Revert "Clear" Button from <button> to <a> Tag for Improved Theme Compatibility while maintaining accessibility improvements

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Change "button" to "a" for theme compatibility

The reset variations button has been converted back to a link with proper accessibility enhancements:
- Added role="button" to the link
- Added keyboard event handling (Enter/Space) to match button behavior
- Added screen reader alert for when variations are reset

This reverts PR #50183 which changed the link to a button, while preserving its accessibility improvements.
The button styling was causing visual issues by making the Clear button more prominent than Add to Cart.

imanish003 committed:
Update tests

imanish003 commented:
@roykho @samueljseay JFYI I've assigned you both as reviewers on the new PR #54320, as I needed to close the previous one (see [comment](https://github.com/woocommerce/woocommerce/pull/54309#issuecomment-2580033199)).

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

roykho commented on plugins/woocommerce/client/legacy/js/frontend/add-to-cart-variation.js:
```diff
@@ -603,6 +606,16 @@
+	 * Handle reset key down event for accessibility.
+	 */
+	VariationForm.prototype.onResetKeyDown = function ( event ) {
+		if ( event.keyCode === 32 || event.keyCode === 13 ) {
```
I would change this to using the `code` property instead `if ( event.code === 'Enter' || event.code === 'Space' ) {`

imanish003 replied:
Good point. I'll update the PR to use `event.code`.

roykho reviewed (changes requested):
I left a comment.

Also do we not need any `aria-label` for the clear link? Another thing I observed is that right after you clear the link with voice over on, you first hear the "select options..." message first before you hear the "clear..." message. Is that expected?

imanish003 committed:
Use event.code instead of event.keyCode

This change follows modern web standards as `keyCode` is deprecated. Also adds JSDoc param documentation for the event parameter.

imanish003 committed:
Accessibility: Add aria-label to Clear variations link

Added an aria-label to provide better context for screen reader users, changing from
just "Clear" to "Clear options" for improved clarity.

imanish003 commented:
Hi @roykho, Thank you for your thorough review. You've raised some important points.

> Also do we not need any aria-label for the clear link?

Good eyes! It makes sense to also add `aria-label` to provide better context for screen reader users, changing from just "Clear" to "Clear options". I have made the changes.

> Another thing I observed is that right after you clear the link with voice over on, you first hear the "select options..." message first before you hear the "clear..." message. Is that expected?

After investigating the [code](https://github.com/woocommerce/woocommerce/blob/d02020dcc8a81075ec362a956bf336217b5af880/plugins/woocommerce/client/legacy/js/frontend/add-to-cart-variation.js#L62-L71), I found that this behavior occurs because `announce_reset` is called after updating the attribute field values. I attempted to resolve it by moving `announce_reset` before the change event trigger, but the issue persisted. While this warrants further investigation, I don't believe we should block this CFE since this is existing behavior rather than a regression. What are your thoughts on proceeding with the current changes? Maybe we can create a follow up issue to fix this?

---

I have addressed your feedback. Can you please review the PR again?

roykho commented on plugins/woocommerce/templates/single-product/add-to-cart/variable.php:
```diff
@@ -45,7 +45,7 @@
 									)
 								);
-								echo end( $attribute_keys ) === $attribute_name ? wp_kses_post( apply_filters( 'woocommerce_reset_variations_link', '<a class="reset_variations" href="#">' . esc_html__( 'Clear', 'woocommerce' ) . '</a>' ) ) : '';
+								echo end( $attribute_keys ) === $attribute_name ? wp_kses_post( apply_filters( 'woocommerce_reset_variations_link', '<a class="reset_variations" href="#" aria-label="' . esc_html__( 'Clear options', 'woocommerce' ) . '">' . esc_html__( 'Clear', 'woocommerce' ) . '</a>' ) ) : '';
```
Because this is an attribute, we should use `esc_attr__()` instead.

roykho reviewed (changes requested):
Thanks for making the changes. I found one small issue.

As for the voice over, yeah please create a separate issue for that.

imanish003 committed:
Replace `esc_html__()` with `esc_attr__()`

We were using `esc_html__()` for the aria-label attribute of the reset variations link. This is incorrect as HTML attributes should use `esc_attr__()` instead.

imanish003 commented:
@roykho Good catch! I have replaced `esc_html__()` with `esc_attr__()` in for `aria-label` attribute. Can you please review the PR again?

PS: I'd appreciate it if we could merge this PR on Monday before the RC release to include it in testing.

CC: @kmanijak @tjcafferkey Could you please review and help merge this PR on Monday? Given the timezone differences with @roykho, waiting for his review may delay inclusion in Monday's release. Thanks! 🙂

roykho reviewed (approved):
Thanks for making the changes and great work! It looks good to me now.

✅ The HTML is now rendering an anchor tag with aria-label
✅ Using voiceover, it correctly announces the "Clear options" selection
✅ Using keyboard, I am able to clear the options with "Space" and "Enter" keys.
✅ After clearing the options, I can hear a message stating my options are cleared.

samueljseay reviewed (approved):
@roykho did a great job of the review on this and I don't have anything to add. I'm sorry I can't jump in and test this one, but had a look over the follow up changes and it looks good to me.

imanish003 committed:
Merge branch 'release/9.6' into fix/revert-clear-button-to-link-with-a11y-2

kmanijak committed:
Merge branch 'release/9.6' into fix/revert-clear-button-to-link-with-a11y-2

kmanijak commented:
[CFE has been approved](https://github.com/woocommerce/woocommerce/issues/54310#issuecomment-2579980221). PR is approved twice, merging on behalf of Manish 🙌
